core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -146,64 +146,65 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
     case Some(topic) => Map("topic" -> topic)
   }
 
+  case class LazyMeter(metricType: String,
+                       eventType: String) {
+    var lazyMeter: Meter = _
+
+    def meter(): Meter = {
+      if (lazyMeter == null)
+        lazyMeter = newMeter(metricType, eventType, TimeUnit.SECONDS, tags)
+      lazyMeter
+    }
+
+    def close(): Unit = {
+      removeMetric(metricType, tags)
+    }
+
+    if (tags.isEmpty) // greedily initialize the general topic metrics
+      meter()
+  }
+
   // an internal map for "lazy initialization" of certain metrics
-  private val metricTypeMap = new Pool[String, Meter]
-
-  def messagesInRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.MessagesInPerSec,
-    newMeter(BrokerTopicStats.MessagesInPerSec, "messages", TimeUnit.SECONDS, tags))
-  def bytesInRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.BytesInPerSec,
-    newMeter(BrokerTopicStats.BytesInPerSec, "bytes", TimeUnit.SECONDS, tags))
-  def bytesOutRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.BytesOutPerSec,
-    newMeter(BrokerTopicStats.BytesOutPerSec, "bytes", TimeUnit.SECONDS, tags))
-  def bytesRejectedRate = metricTypeMap.getAndMaybePut(BrokerTopicStats.BytesRejectedPerSec,
-    newMeter(BrokerTopicStats.BytesRejectedPerSec, "bytes", TimeUnit.SECONDS, tags))
+  // public for testing only
+  val metricTypeMap = mutable.Map[String, LazyMeter](
```

Comment:
This has to be thread-safe, right? Same for LazyMeter and `removeMetricsHelper`.

Reply:
Great catch. I _think_ the newest implementation should be thread-safe, although it is late so I may have an error somewhere